specs/agents/metadata.md

```diff
@@ -99,6 +99,111 @@ metadata is available.
+
+As an example with curl, first, an API token must be created
+
+```sh
+TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 300"`
+```
+
+Then, metadata can be retrieved, passing the API token
+
+```sh
+curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data
+```
+
+From the returned metadata, the following fields are useful
+
+| Cloud metadata field  | AWS Metadata field  |
+| --------------------  | ------------------- |
+| `account.id`          | `accountId`         |
+| `instance.id`         | `instanceId`        |
+| `availability_zone`   | `availabilityZone`  |
+| `machine.type`        | `instanceType`      |
+| `provider`            | aws                 |
+| `region`              | `region`            |
+
+#### GCP metadata
+
+Metadata about a GCP machine instance can be retrieved from the 
+metadata service, `http://metadata.google.internal`.
+
+An example with curl
+
+```sh
+curl -X GET "http://metadata.google.internal/computeMetadata/v1/?recursive=true" -H "Metadata-Flavor: Google"
+```
+
+From the returned metadata, the following fields are useful
+
+| Cloud metadata field  | AWS Metadata field  |
+| --------------------  | ------------------- |
+| `instance.id`         | `instance.id`       |
+| `instance.name`       | `instance.name`     |
+| `project.id`          | `project.numericProjectId` as a string |
+| `project.name`        | `project.projectId` |
+| `availability_zone`   | last part of `instance.zone`, split by `/`  |
+| `machine.type`        | last part of `instance.machineType`, split by `/` |
+| `provider`            | gcp                 |
+| `region`              | last part of `instance.zone`, split by `-`            |
+
+#### Azure metadata
+
+##### Azure VMs
+
+Metadata about an Azure VM can be retrieved from the internal metadata
+endpoint, `http://169.254.169.254`.
+
+An example with curl
+
+```sh
+curl -X GET "http://169.254.169.254/metadata/instance/compute?api-version=2019-08-15" -H "Metadata: true"
+```
+
+From the returned metadata, the following fields are useful
+
+| Cloud metadata field  | AWS Metadata field  |
+| --------------------  | ------------------- |
+| `account.id`          | `subscriptionId`    |
+| `instance.id`         | `vmId`              |
+| `instance.name`       | `name`              |
+| `project.name`        | `resourceGroupName` |
+| `availability_zone`   | `zone`              |
+| `machine.type`        | `vmSize`            |
+| `provider`            | azure               |
+| `region`              | `location`          |
+
+##### Azure App Services _(Optional)_
+
+Azure App Services are a PaaS offering within Azure which does not
+have access to the internal metadata endpoint. Metadata about
+an App Service can however be retrieved from environment variables
+
```

Comment:
@russcam  (or anyone) -- do you think we need to define behavior for what-to-do when one of the environment variables is not present or has an unexpected form?  Unlike the instance endpoints these variables aren't a formal API and I have vague concerns about what to do when/if something isn't set.Put another way -- if an variable like `WEBSITE_RESOURCE_GROUP` changes its format or a variable like `WEBSITE_INSTANCE_ID` is no longer set, is it better to _not_ report this data, reporting some constant string like `UNKNOWN`, or just not report the data.  

My knee jerk response in setting something like `UNKNOWN` -- but that's without thinking about too much.

Reply:
Very good question! In the case that an environment variable is missing or the value is in an unexpected format, I think it's reasonable to not report the cloud metadata. That's what I'm currently doing in the .NET implementation. 

Longer term, I think we'd want to have automated tests for cloud metatdata as part of the larger work of cloud services integrations. Hopefully, these automated tests would give us warning if the format changes.

Reply:
The gherkin specs added in https://github.com/elastic/apm/pull/365/commits/97680396a8e2ddd9ec169a3fc617491423f49caf have scenarios for missing env vars or unexpected formats.